core/server/src/main/java/alluxio/worker/block/BlockMasterSync.java

```diff
@@ -177,9 +186,21 @@ private void handleMasterCommand(Command cmd) throws Exception {
         break;
       // Master requests blocks to be removed from Alluxio managed space.
       case Free:
-        for (long block : cmd.getData()) {
-          mBlockRemovalService.execute(new BlockRemover(mBlockWorker,
-              Sessions.MASTER_COMMAND_SESSION_ID, block));
+        synchronized (mRemovingBlockIdToFinished) {
+          for (long block : cmd.getData()) {
+            if (!mRemovingBlockIdToFinished.containsKey(block)) {
+              mRemovingBlockIdToFinished.put(block, false);
+              mBlockRemovalService.execute(new BlockRemover(mBlockWorker,
+                  mRemovingBlockIdToFinished, Sessions.MASTER_COMMAND_SESSION_ID, block));
+            }
+          }
+          Iterator<Map.Entry<Long, Boolean>> it = mRemovingBlockIdToFinished.entrySet().iterator();
```

Comment:
If it is possible to use a set instead of a map, I think the removal from the set can be done by the removal thread.